numba/targets/arrayobj.py

```diff
@@ -979,6 +1052,141 @@ def fancy_getitem_array(context, builder, sig, args):
+    else:
+        # Non-contiguous array: need to examine strides
+        lower = zero
+        upper = zero
+        for i in range(arrty.ndim):
+            # Compute the largest byte offset on this dimension
+            #   max_axis_offset = strides[i] * (shapes[i] - 1)
+            # (shapes[i] == 0 is catered for by the empty array case below)
+            max_axis_offset = builder.mul(strides[i],
+                                          builder.sub(shapes[i], one))
+            is_upwards = builder.icmp_signed('>=', max_axis_offset, zero)
+            # Expand either upwards or downwards depending on stride
+            upper = builder.select(is_upwards,
+                                   builder.add(upper, max_axis_offset), upper)
+            lower = builder.select(is_upwards,
+                                   lower, builder.add(lower, max_axis_offset))
+        # Return a half-open range
+        upper = builder.add(upper, itemsize)
+        # Adjust for empty arrays
+        is_empty = builder.icmp_signed('==', arr.nitems, zero)
+        upper = builder.select(is_empty, zero, upper)
+        lower = builder.select(is_empty, zero, lower)
+
+    return lower, upper
+
+
+def compute_memory_extents(context, builder, lower, upper, data):
+    """
+    Given [lower, upper) byte offsets and a base data pointer,
+    compute the memory pointer bounds as pointer-sized integers.
+    """
+    data_ptr_as_int = builder.ptrtoint(data, lower.type)
+    start = builder.add(data_ptr_as_int, lower)
+    end = builder.add(data_ptr_as_int, upper)
+    return start, end
+
+def get_array_memory_extents(context, builder, arrty, arr, shapes, strides, data):
+    """
+    Compute a half-open range [start, end) of pointer-sized integers
+    which fully contain the array data.
+    """
+    lower, upper = offset_bounds_from_strides(context, builder, arrty, arr,
+                                              shapes, strides)
+    return compute_memory_extents(context, builder, lower, upper, data)
+
+
+def extents_may_overlap(context, builder, a_start, a_end, b_start, b_end):
+    """
+    Whether two memory extents [a_start, a_end) and [b_start, b_end)
+    may overlap.
+    """
+    # Comparisons are unsigned, since we are really comparing pointers
+    may_overlap = builder.and_(
+        builder.icmp_unsigned('<', a_start, b_end),
+        builder.icmp_unsigned('<', b_start, a_end),
+        )
+    return may_overlap
+
+
+def maybe_copy_source(context, builder, use_copy,
+                      srcty, src, src_shapes, src_strides, src_data):
+    ptrty = src_data.type
+
+    copy_layout = 'C'
+    copy_data = cgutils.alloca_once_value(builder, src_data)
+    copy_shapes = src_shapes
+    copy_strides = None  # unneeded for contiguous arrays
+
+    with builder.if_then(use_copy, likely=False):
+        # Allocate temporary scratchpad
+        # XXX: should we use a stack-allocated array for very small
+        # data sizes?
```

Comment:
Instead of manually customizing the codegen for small stack allocated arrays, directly exposing malloc/free may allow LLVM passes to do that automatically.  I have seen that work for small cases.